lib/ContextResolver.js

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright (c) 2019 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+const {
+  isArray: _isArray,
+  isObject: _isObject,
+  isString: _isString,
+} = require('./types');
+const {prependBase} = require('./url');
+const JsonLdError = require('./JsonLdError');
+const ResolvedContext = require('./ResolvedContext');
+
+const MAX_CONTEXT_URLS = 10;
+
+module.exports = class ContextResolver {
+  /**
+   * Creates a ContextResolver.
+   *
+   * @param sharedCache a shared LRU cache with `get` and `set` APIs.
+   */
+  constructor({sharedCache}) {
+    this.perOpCache = new Map();
+    this.sharedCache = sharedCache;
+  }
+
+  async resolve({context, documentLoader, base, cycles = new Set()}) {
+    // process `@context`
+    if(context && _isObject(context) && context['@context']) {
+      context = context['@context'];
+    }
+
+    // context is one or more contexts
+    if(!_isArray(context)) {
+      context = [context];
+    }
+
+    // resolve each context in the array
+    const allResolved = [];
+    for(const ctx of context) {
+      if(_isString(ctx)) {
+        // see if `ctx` has been resolved before...
+        let resolved = this._get(ctx);
+        if(!resolved) {
+          // not resolved yet, resolve
+          resolved = await this._resolveRemoteContext(
+            {url: ctx, documentLoader, base, cycles});
```

Comment:
Does this have the affect of serializing all remote context fetches?

Reply:
`_resolveRemoteContext` will fetch all remote contexts (recursively) and return an array of resolved contexts. Note that this does not include any scoped contexts, i.e., it doesn't deeply inspect the resolved contexts for those, but it will resolve any relative URLs encountered in scoped contexts to ensure the base URL used is proper. Any scoped contexts will be resolved later, JIT.

Reply:
If by `this` you meant "the `await`" -- then, yes, an array of context URLs will be loaded serially now. A future PR can add more complexity here to parallelize that.